src/Stratis.Bitcoin.Features.RPC/RPCClient.cs

```diff
@@ -292,6 +329,106 @@ private string GetCookiePath()
             return null;
         }
 
+        /// <summary>
+        /// The RPC Capabilities of this RPCClient instance, this property will be set by a call to ScanRPCCapabilitiesAsync
+        /// </summary>
+        public RPCCapabilities Capabilities { get; set; }
+
+        /// <summary>
+        /// Run several RPC function to scan the RPC capabilities, then set RPCClient.Capabilities
+        /// </summary>
+        /// <returns>The RPCCapabilities</returns>
+        public async Task<RPCCapabilities> ScanRPCCapabilitiesAsync()
+        {
+            var capabilities = new RPCCapabilities();
+            var rpc = this.PrepareBatch();
+            var waiting = Task.WhenAll(
+            SetVersionAsync(capabilities),
+            CheckCapabilities(rpc, "scantxoutset", v => capabilities.SupportScanUTXOSet = v),
+            CheckCapabilities(rpc, "signrawtransactionwithkey", v => capabilities.SupportSignRawTransactionWith = v),
+            CheckCapabilities(rpc, "estimatesmartfee", v => capabilities.SupportEstimateSmartFee = v),
+            CheckCapabilities(rpc, "generatetoaddress", v => capabilities.SupportGenerateToAddress = v),
+            CheckSegwitCapabilitiesAsync(rpc, v => capabilities.SupportSegwit = v));
+            await rpc.SendBatchAsync();
+            await waiting;
+            Thread.MemoryBarrier();
+
+            this.Capabilities = capabilities;
+            return capabilities;
+        }
+
+        private async Task SetVersionAsync(RPCCapabilities capabilities)
+        {
+            try
+            {
+                var getInfo = await SendCommandAsync(RPCOperations.getnetworkinfo);
+                capabilities.Version = ((JObject)getInfo.Result)["version"].Value<int>();
+                capabilities.SupportGetNetworkInfo = true;
+                return;
+            }
+            catch (RPCException ex) when (ex.RPCCode == RPCErrorCode.RPC_METHOD_NOT_FOUND || ex.RPCCode == RPCErrorCode.RPC_METHOD_DEPRECATED)
+            {
+                capabilities.SupportGetNetworkInfo = false;
+            }
+
+            {
+#pragma warning disable CS0618 // Type or member is obsolete
+                var getInfo = await SendCommandAsync(RPCOperations.getinfo);
+#pragma warning restore CS0618 // Type or member is obsolete
+                capabilities.Version = ((JObject)getInfo.Result)["version"].Value<int>();
+            }
+        }
+
+        /// <summary>
+        /// Run several RPC function to scan the RPC capabilities, then set RPCClient.RPCCapabilities
+        /// </summary>
+        /// <returns>The RPCCapabilities</returns>
+        public RPCCapabilities ScanRPCCapabilities()
+        {
+            return ScanRPCCapabilitiesAsync().GetAwaiter().GetResult();
+        }
+
+        private static async Task CheckSegwitCapabilitiesAsync(RPCClient rpc, Action<bool> setResult)
```

Comment:
Obviously you've looked at this in a lot more detail than I have, but is this really the correct way to determine segwit capability? The validity of an address does not, in my opinion, indicate underlying protocol support. Particularly in light of the soft fork nature of segwit and how 'legacy' nodes can still process the transactions in blocks, albeit in a limited way.

Reply:
I'm not sure if this the correct way but it is the way that NBitcoin is checking the nodes capabilities:

https://github.com/MetacoSA/NBitcoin/blob/master/NBitcoin/RPC/RPCClient.cs#L370-#L388

Reply:
Further, the validity would need to be network-specific too

Reply:
> I'm not sure if this the correct way but it is the way that NBitcoin is checking the nodes capabilities:
> 
> https://github.com/MetacoSA/NBitcoin/blob/master/NBitcoin/RPC/RPCClient.cs#L370-#L388

Ok, fair enough. Let me think about this a bit